File: template-method/cpp/compiler.cpp
#ifndef ___COMPILER___
#define ___COMPILER___

#include <string>
#include <iostream>

class Compiler {

public:

    void build(std::string code) {
		std::cout << "Source code: " << code << std::endl;
		code = preprocess(code);
		code = compile(code);
		code = assemble(code);
		code = link(code);
	}

private:

    virtual std::string preprocess(const std::string &code) = 0;
    virtual std::string compile(const std::string &code) = 0;
    virtual std::string assemble(const std::string &code) = 0;
    virtual std::string link(const std::string &code) = 0;
};

#endif // ___COMPILER___

File: template-method/dart/compiler.dart
abstract class Compiler {

    void build(String code) {
		print("Source code: $code\n");
		code = preprocess(code);
		code = compile(code);
		code = assemble(code);
		code = link(code);
	}

    String preprocess(String code);
    String compile(String code);
    String assemble(String code);
    String link(String code);
}
